packages/core/js-client/src/api.ts

```diff
@@ -15,182 +15,158 @@
-  // TODO: Support this in aqua-to-js package
-  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
-  const service: ServiceImpl = args.pop() as ServiceImpl;
+  const functionArgs: Record<string, SimpleTypes | ArrowWithoutCallbacks> =
+    def.arrow.domain.tag === "nil" ? {} : def.arrow.domain.fields;
 
-  const { peer, serviceId } = extractServiceArgs(args, def.defaultServiceId);
+  let peer: FluencePeer | undefined;
 
-  registerService({
-    def,
-    service,
-    serviceId,
-    peer,
-  });
-};
-
-function isConfig(arg: unknown): arg is FnConfig {
-  return typeof arg === "object" && arg !== null;
-}
-
-/**
- * Arguments could be passed in one these configurations:
- * [...actualArgs]
- * [peer, ...actualArgs]
- * [...actualArgs, config]
- * [peer, ...actualArgs, config]
- *
- * This function select the appropriate configuration and returns
- * arguments in a structured way of: { peer, config, args }
- */
-function extractFunctionArgs(
-  args: unknown[],
-  def: FunctionCallDef,
-): {
-  client: FluencePeer;
-  config: FnConfig;
-  args: PassedArgs;
-} {
-  const argumentTypes = getArgumentTypes(def);
-  const argumentNames = Object.keys(argumentTypes);
-  const numberOfExpectedArgs = argumentNames.length;
-
-  let peer: FluencePeer;
-  let config: FnConfig;
-
-  if (isFluencePeer(args[0])) {
+  if (args[0] instanceof FluencePeer) {
     peer = args[0];
     args = args.slice(1);
   } else {
-    if (Fluence.defaultClient == null) {
-      throw new Error(
-        "Could not register Aqua service because the client is not initialized. Did you forget to call Fluence.connect()?",
-      );
-    }
-
     peer = Fluence.defaultClient;
   }
 
-  const maybeConfig = args[numberOfExpectedArgs];
-
-  if (isConfig(maybeConfig)) {
-    config = maybeConfig;
-  } else {
-    config = {};
-  }
-
-  const structuredArgs = args.slice(0, numberOfExpectedArgs);
+  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+  const config =
+    argCount < args.length
+      ? // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+        (args.pop() as CallAquaFunctionConfig | undefined)
+      : undefined;
 
-  if (structuredArgs.length !== numberOfExpectedArgs) {
+  if (peer == null) {
     throw new Error(
-      `Incorrect number of arguments. Expecting ${numberOfExpectedArgs}`,
+      "Could not register Aqua service because the client is not initialized. Did you forget to call Fluence.connect()?",
     );
   }
 
-  const argsRes = argumentNames.reduce((acc, name, index) => {
-    return { ...acc, [name]: structuredArgs[index] };
-  }, {});
+  const callArgs = Object.fromEntries<
+    JSONValue | ((...args: JSONValue[]) => JSONValue)
+  >(
+    args.slice(0, argCount).map((arg, i) => {
+      const argSchema = functionArgs[argNames[i]];
+
+      if (argSchema.tag === "arrow") {
+        if (typeof arg !== "function") {
+          throw new Error("Argument and schema doesn't match");
```

Comment:
```suggestion
          throw new Error("Argument and schema don't match");
```

Don't match how, why? What to do with such an error, is not clear. Let's make the error actionable, add reason there and what the user should check.